packages/nextjs/hooks/scaffold-eth/useScaffoldEventHistory.ts

```diff
@@ -45,135 +80,82 @@ export const useScaffoldEventHistory = <
-        }
-        setError(e);
-        console.error(e);
-      } finally {
-        setIsLoading(false);
-      }
-    },
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [
-      blockData,
-      deployedContractData,
-      enabled,
-      eventName,
-      events,
-      // eslint-disable-next-line react-hooks/exhaustive-deps
-      JSON.stringify(filters, replacer),
-      fromBlockUpdated,
-      publicClient,
-      receiptData,
-      transactionData,
+  const [isFirstRender, setIsFirstRender] = useState(true);
+
+  const { data: blockNumber } = useBlockNumber({ watch: watch, chainId: targetNetwork.id });
+
+  const { data: deployedContractData } = useDeployedContractInfo(contractName);
+
+  const event =
+    deployedContractData &&
+    ((deployedContractData.abi as Abi).find(part => part.type === "event" && part.name === eventName) as AbiEvent);
+
+  const query = useInfiniteQuery({
+    queryKey: [
+      "eventHistory",
+      {
+        contractName,
+        address: deployedContractData?.address,
+        eventName,
+        fromBlock: fromBlock.toString(),
+        chainId: targetNetwork.id,
+      },
     ],
-  );
-
-  useEffect(() => {
-    if (!deployedContractLoading) {
-      readEvents();
-    }
-  }, [readEvents, deployedContractLoading]);
-
-  useEffect(() => {
-    // Reset the internal state when target network or fromBlock changed
-    setEvents([]);
-    setFromBlockUpdated(fromBlock);
-    setError(undefined);
-  }, [fromBlock, targetNetwork.id]);
-
-  useInterval(
-    async () => {
-      if (!deployedContractLoading) {
-        readEvents();
-      }
+    queryFn: async ({ pageParam }) => {
+      if (!Boolean(deployedContractData?.address) || !Boolean(publicClient)) return undefined;
+      const data = await getEvents(
+        { address: deployedContractData?.address, event, fromBlock: pageParam, args: filters },
+        publicClient,
+        { blockData, transactionData, receiptData },
+      );
+
+      return data;
     },
-    watch && enabled ? (targetNetwork.id !== chains.hardhat.id ? scaffoldConfig.pollingInterval : 4_000) : null,
-  );
-
-  const eventHistoryData = useMemo(
-    () =>
-      events?.map(addIndexedArgsToEvent) as UseScaffoldEventHistoryData<
+    enabled: enabled && Boolean(deployedContractData?.address) && Boolean(publicClient),
+    initialPageParam: fromBlock,
+    getNextPageParam: () => {
+      return blockNumber;
+    },
+    select: data => {
+      const events = data.pages.flat();
+      const eventHistoryData = events?.map(addIndexedArgsToEvent) as UseScaffoldEventHistoryData<
         TContractName,
         TEventName,
         TBlockData,
         TTransactionData,
         TReceiptData
-      >,
-    [events],
-  );
+      >;
+      return {
+        pages: eventHistoryData?.reverse(),
+        pageParams: data.pageParams,
+      };
+    },
+  });
+
+  useEffect(() => {
+    const shouldSkipEffect = !blockNumber || !watch || isFirstRender;
```

Comment:
Maybe you can remove **watch** from here?

Reply:
Added `watch` to `useEffect` dependency, this will allow people to programatically disable `watch` mode in their app after certain time / interactions, checkout https://github.com/scaffold-eth/scaffold-eth-2/pull/869#discussion_r1651362844 🙌